### ISSUE TITLE
Python: Make tests pass against shared test data

### ISSUE THREAD

**Issue by aslakhellesoy** [cucumber/tag-expressions#18]
In #9 we introduced shared test data. The Python implementation should be updated to:

* [ ] Run all tests in `testdata/*.yml`
* [ ] Update the implementations to make those tests pass

**Comment by jenisys** [comment]
Pull #95 contains an implementation to use the test data for Python.

* [x] Run all tests in `testdata/*.yml`
* [x] Update the implementations to make those tests pass

HINTS:

* The backslash-escaping mechanism is not implemented yet for Python
* Therefore, tests that use the backslash-escaping mechanism are currently skipped

**Comment by luke-hill** [comment]
@jenisys - Not sure if you're still around / maintaining and checking python. But 2/3 of the shared tests are now ran. Just the errors ones aren't running. This feels like this issue is close to being "close off-able"

cc/ @brasmusson

**Comment by kieran-ryan** [comment]
> ...2/3 of the shared tests are now ran. Just the errors ones aren't running. This feels like this issue is close to being "close off-able"

@luke-hill, interested in your thoughts based on diff below. The Python error messages have higher granularity; so we may want to consolidate a similar approach into the other parsers; and/or limit the expected message to a suitable subset and check that it is contained within error messages of each implementation.

Some [error messages improvements in recent Python releases](https://docs.python.org/3/whatsnew/3.13.html#improved-error-messages) which may be of relevance regards enhancements.

<details>
<summary>Python error messages diff</summary>
<b>Note:</b> `cucumber_tag_expressions.parser.TagExpressionError: ` dropped from expected error messages.

#### @a @b or

```diff
- Tag expression "@a @b or" could not be parsed because of syntax error: Expected operator.
+ Syntax error. Expected operator after @a
+ Expression: @a @b or
+ _______________^ (HERE)
```

#### @a and (@b not)

```diff
- Tag expression "@a and (@b not)" could not be parsed because of syntax error: Expected operator.
+ Syntax error. Expected operator after @b
+ Expression: @a and ( @b not )
+ ________________________^ (HERE)
```

#### @a and (@b @c) or

```diff
- Tag expression "@a and (@b @c) or" could not be parsed because of syntax error: Expected operator.
+ Syntax error. Expected operator after @b
+ Expression: @a and ( @b @c ) or
+ ________________________^ (HERE)
```

#### @a and or

```diff
- @a and or" could not be parsed because of syntax error: Expected operand.
+ Syntax error. Expected operand after and
+ Expression: @a and or
+ ___________________^ (HERE)
```

#### or or

```diff
- Tag expression "or or" could not be parsed because of syntax error: Expected operand.
+ Syntax error. Expected operand after BEGIN
+ Expression: or or
+ _____________^ (HERE)
```

#### a and or

```diff
- Tag expression "a and or" could not be parsed because of syntax error: Expected operand.
+ Syntax error. Expected operand after and
+ Expression: a and or
+ __________________^ (HERE)
```

#### a b

```diff
- Tag expression "a b" could not be parsed because of syntax error: Expected operator.
+ Syntax error. Expected operator after a
+ Expression: a b
+ ______________^ (HERE)
```

#### ( a and b ) )

```diff
- Tag expression "( a and b ) )" could not be parsed because of syntax error: Unmatched ).
+ Missing '(': Too few open-parens in: ( a and b ) )
+ Expression: ( a and b ) )
+ ________________________^ (HERE)
```

#### x\ or y

```diff
- Tag expression "x\\ or y" could not be parsed because of syntax error: Expected operator.
+ Syntax error. Expected operator after x or
+ Expression: x or y
+ _________________^ (HERE)
```

#### ( ( a and b )

```diff
- Tag expression "( ( a and b )" could not be parsed because of syntax error: Unmatched (
+ Unclosed '(': Too many open-parens in: ( ( a and b )
```
</details>

**Comment by luke-hill** [comment]
This should come from a consensus at one of the contributor meetings.

Ideally the shared ones should conform. I think the issue here is development came before testing, which ideally shouldn't happen. But such is the way with real life. 

P.S. Looking at some of the above. I think there are some errors in the messages.

`or or` => The `^` marker is 1 char too far right
`( a and b ) )` => The previous error probably was more informative than this one now.
`x\ or y` => seems to be sanitizing out the `\` char
`( ( a and b )` same as the opposite one.

**Comment by kieran-ryan** [comment]
Cheers @luke-hill! 🙌 I might take a look at correcting the error messages as you've mentioned if I get chance - was interested to look at enhancing and maybe taking a similar format to the [latest Python error messages style](https://docs.python.org/3/whatsnew/3.13.html#improved-error-messages) - good to have direction

**Comment by mpkorstanje** [comment]
I'm definitely in favor of friendlier messages! Cucumber expressions currently uses the format below. Which contains the same elements as the Python error messages but with a slightly different formatting. 

```
expression: three (brown)/black mice
text: three brown mice
exception: |-
  This Cucumber Expression has a problem at column 7:

  three (brown)/black mice
        ^-----^
  An alternative may not exclusively contain optionals.
  If you did not mean to use an optional you can use '\(' to escape the '('
```  

But it seems prudent to me to make sure the existing tests work before we start making improvements. 

@kieran-ryan I don't quite understand your diff. I don't see any code in the current Python implementation that would produce the error messages you are showing.

**Comment by kieran-ryan** [comment]
> @kieran-ryan I don't quite understand your diff. I don't see any code in the current Python implementation that would produce the error messages you are showing. Is this a proposal for improvement or how it currently works? If the latter, could you point out to me where in the code this is done?

@mpkorstanje, above diff would be how it currently works rather than a proposal. I've [enabled the error message tests to run against the Python implementation](https://github.com/cucumber/tag-expressions/actions/runs/12753687341/job/35545886484?pr=184) - to show same, though the pytest output is not most the most user-friendly to read; may find useful in any case.

The code handling the error message is as below. Based on internal usage within the library, you can ignore the first conditional (`error_index > len(parts)`) - as _believe_ is never hit; and for awareness, the second conditional is always hit (`if message`).

https://github.com/cucumber/tag-expressions/blob/0ddbd2f15f49b6d4a29f1924acb31b3972dc3f1f/python/cucumber_tag_expressions/parser.py#L324-L333

**Comment by kieran-ryan** [comment]
@mpkorstanje, @luke-hill I aligned the Python error messages on the attached pull request; with tests passing. Pull request 'changes' will give a good idea of the difference - though bear in mind that most tests are checking for a substring, so although appears information has been dropped in some cases, other parts of the error message should majorly retain that information - such as the `____^ (HERE)` aspect. Interested in your perspective whether we should proceed with these error messages; or refactor any/all messages (particularly if any are considered to be a regression).

**Comment by luke-hill** [comment]
I'm definitely "more" pro than con. So from me it's a :+1: but I would prefer that we "didn't" do this for python originally and proposed it for all flavours.

But I think that because python is a bit like Reqnroll where it's got a lot of custom components from pytest-bdd e.t.c. and it's not fully using a cucumber official implementation, we're likely to hit a few of these.

I think the tag expressions library got a lot of it's juice early in the inception stage/s, so giving them a facelift and an update I see only as a positive.

TL;DR - I like the updates, but would prefer any changes are fully unanimous across all flavours of tag-expressions

**Comment by luke-hill** [comment]
Thinking out loud too. To make things easier, is it possible we could take the following development path @kieran-ryan 

1) Standardise python against the existing error messages (If not a lot of work)
2) Re-enable the full 3 yml files to be ran, parsed and validated against
---- Cut version here ----
3) Make the requisite changes to the YML to facilitate the new python error messages **AND** cherry-pick the code changes to make this pass into a single branch / PR - So we can see the new code changes against the new yml file (TBD)
4) With this PR change visible, we then co-ordinate other maintainers to come in / you to go in and fix the various flavours

WDYT>?
